### PR TITLE
Support for NumericLiteralSeparator, Stage 1 feature

### LINKED ISSUE #5783: Numeric Separator: Stage 1
Original issue submitted by @hzoo in https://github.com/babel/babylon/issues/538

> Advance Numeric separators proposal to Stage 1 slides (Sam Goto) (@samuelgoto)
cc @ajklein and @domenic

## Info
Proposed at TC39 Meeting: [May 2017](https://github.com/tc39/agendas/blob/master/2017/05.md)
Slides at the meeting: https://docs.google.com/presentation/d/1hcajTemZB2Ruo4EePOyFiva1xpyv5ukKk4aQ0B83dUA/edit#slide=id.p
Proposal Repo: https://github.com/samuelgoto/proposal-numeric-separator

## Example
```js
var thousands = 10_000; // Instead of 10000.
var credit_card_number = 1234_5678_9012_3456; // Instead of 123456789012345.
var social_security_number = 999_99_9999; // Instead of 999999999.
var pi = 3.14_15; // Instead of 3.1415
var bytes = 0b11010010_01101001_10010100_10010010; // Instead of 0b11010010011010011001010010010010.
var 0xCAFE_F00D; // Instead of 0XCAFEF00D.
```

## Basic Rules
- to use the `_` character.
- only one consecutive underscore is allowed.
- only between digits (not allowed at the beginning or end of literals)

## Parser

> Done in https://github.com/babel/babylon/pull/541 by @rwaldron

Plugin: `numericSeparator`
```js
0b0_1_1;
```

```json
{
  "type": "NumericLiteral",
  "extra": {
    "rawValue": 3,
    "raw": "0b0_1_1"
  },
  "value": 3
}
```

## Transform
- if the AST is simple (just the value/raw) would be trivial to remove

https://github.com/babel/babel/wiki/Adding-a-new-Proposal-to-Babel

### PR DESCRIPTION
**Pull request by rwaldron**
~~**Not ready for merge**~~

Signed-off-by: Rick Waldron <waldron.rick@gmail.com>

| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          | no
| Major: Breaking Change?  | no
| Minor: New Feature?      | yes
| Deprecations?            | no
| Spec Compliancy?         | yes
| Tests Added/Pass?        | yes
| Fixed Tickets            | Fixes #5783
| License                  | MIT
| Doc PR                   | 
| Dependency Changes       |  yes

Adds support for Numeric Separator

### CODE REVIEW

**Comment by rwaldron** [comment]
@hzoo ready for an initial review!

File: packages/babel-plugin-syntax-numeric-separator/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# babel-plugin-syntax-numeric-separator
+
+
```
Comment (xtuc) [inline packages/babel-plugin-syntax-numeric-separator/README.md]: To be consistent across package READMEs, could you add there a description:

```md
# babel-plugin-syntax-numeric-separator

> Transforms blablabla

## Installation
[...]
```

We should maybe add this to the spec implementor documentation.

**Comment by xtuc** [comment]
Nice job @rwaldron.

Do we want to backport that to 6.x?

**Comment by hzoo** [comment]
I'd like to not have to backport features, especially if it's a proposal?

File: packages/babel-plugin-syntax-numeric-separator/README.md (hunk 1 above)
Comment (rwaldron) [inline packages/babel-plugin-syntax-numeric-separator/README.md]: Yes, absolutely

**Comment by rwaldron** [comment]
@xtuc PR updated with amended commit; now includes: https://github.com/babel/babel/pull/5793/files#diff-921c1523bfcb64eefcc009dba96e1a3bR3

**Comment by xtuc** [comment]
Thanks @rwaldron. I just nit'ed https://github.com/babel/babel/pull/5793/commits/cd8f6e062876261a334d968f875e477a48927b6d, you might want to squash it

File: packages/babel-plugin-syntax-numeric-separator/README.md [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # babel-plugin-syntax-numeric-separator
 
-Allow parsing of Numeric Literals (Decimal, Binary, Hex and Octal) that contain a _NumericLiteralSeparator_.
+> Allow parsing of Numeric Literals (Decimal, Binary, Hex and Octal) that contain a _NumericLiteralSeparator_.
```
Comment (rwaldron) [inline packages/babel-plugin-syntax-numeric-separator/README.md]: Already done :D

Comment (rwaldron) [inline packages/babel-plugin-syntax-numeric-separator/README.md]: https://github.com/babel/babel/pull/5798

**Comment by hzoo** [comment]
@rwaldron I assume the test failures are because it doesn't have your patch yet? I published 7.0.0-beta.12 yesterday so just have to update the 3 deps to use it - can just grep for `"babylon": `

**Comment by rwaldron** [comment]
@hzoo all deps updated to 7.0.0-beta.12, and patch rebased

File: packages/babel-plugin-transform-numeric-separator/README.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+});
+```
+
+## Options
+
+### `toDecimal`
+
+`boolean`, defaults to `false`.
+
+By default, this plugin will only remove the separators, printing Decimal, Binary, Hex and Octal number representations. Enabling this option will  convert them all to a Decimal value.
```
Comment (hzoo) [inline packages/babel-plugin-transform-numeric-separator/README.md]: extra space here?

We do have a http://babeljs.io/docs/plugins/transform-es2015-literals/ to deal with this already so in the ideal case this plugin would defer to that for the transformation. Not sure how that fits but interesting in relation to https://github.com/babel/babel/pull/5735. I'm ok with doing this atm though.

Comment (rwaldron) [inline packages/babel-plugin-transform-numeric-separator/README.md]: This section has been removed.

**Comment by hzoo** [comment]
Thanks a lot @rwaldron! (had to fix a failing issue but CI is being weird, it's https://github.com/babel/babel/commit/1d7d090bf035eeef74db3aafbb9af42d93ca6b9d)

**Pull request merged**